### DISCUSSION TITLE
Welcome to NanoEls Discussions!

### DISCUSSION THREAD

**Discussion by kachurovskiy** [kachurovskiy/nanoels#1] [General]
We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.

## If you're having trouble with your NanoEls

  * Try updating to the latest version of the [code](https://github.com/kachurovskiy/nanoels/blob/main/h1/NanoEls.ino)
  * Try isolating the problem to either encoder, stepper or the PCB
    * To check if encoder is working, uncomment the `// #define SHOW_ANGLE` line in the code, upload it to your Arduino, hand-turn the spindle and check the angle shown on screen
    * To check if stepper is working, use `LEFT` and `RIGHT` buttons, check actual stepper movement and carriage position shown on screen
    * To check if PCB is working, use [multimeter](https://www.google.com/search?client=firefox-b-d&q=multimeter)
      * Check that voltage has expected values (5v / 0v) on 5V / GND pins, 24V on stepper driver power supply
      * Check connectivity between Arduino pins and outputs they are supposed to be connected with e.g. `D2` to `encoder A`
 * If you can't figure it out, please ask here
   * Make sure to provide photos of your setup showing wiring and PCB
   * Describe step-by-step what you're doing / which buttons pressing
   * State which software version you're reporting the problem in (shown if `ON` button is held for `6s`)

## Common problems

 * Wrong pin numbers specified in the Arduino code
   * Special note: encoder pins can't be moved from `D2` and `D3` to any other pins since Arduino Nano only supports interrupts on those 2 pins
 * If encoder readings are sometimes wrong, try adding `2kOhm` resistors between `encoder A` and `5V`, `encoder B` and `5V`

**Comment by Dreher-V8** [comment]
Hello, thanks for your work. It's a nice job, that you do.
I test it on my Lathe, I have written a own Soft to move the Support, but not for threading.
But sun and shadows are in de soft:
V5: all thinks works contrary since V5 vs V3
V6: move to the rightsite in the stopposition. And now it will not start move for thread. I move it al little bit to left an now it works. But   sometimes it moves in the wrong direction to the stop and movs then correct. Also sometimes it moves to fast.
V7: Steppermotor not works.

All DMxxx Driver can't 200 Steps, only 400 and more!
Only the TB6600 Driver can 200 Steps


good new year
Wolfgang

**Reply by kachurovskiy** [reply]
Wolfgang, DM drivers can do full steps, but it might not be marked in the step table. Try placing all SW5-SW8 switches in ON position.

Sorry to hear that something doesn't work. Let's only talk about the last software version. Can you please post some photos of your setup and provide step-by-step description of the problem? Thank you.

**Reply by misterixxxx** [reply]
"На токарном станке с ЧПУ движение суппорта по оси Z к шпинделю (к патрону) обозначается знаком минус (–Z). Это стандартное правило для перемещения инструмента в сторону уменьшения координат.".
На своем станке вижу, что при выборе в верхней строке "минус", каретка идет вправо, от шпинделя, а при "+" идет в сторону шпинделя (по  умолчанию в прошивке). При этом по "стрелкам" работает корректно - левая влево, а правая - вправо. Так же и по РГИ. 
Что нужно изменить в прошивке, чтобы работало по стандарту?

**Comment by Dreher-V8** [comment]
hi,

thanks for the quickly answer.
I will test it and give answer

Setup:
NEMA23 23HS2442B 4.2A
https://www.ebay.de/itm/1PC-Nema23-Schrittmotor-23HS2442B-4-2A-112mm-425oz-in-3Nm-Dual-Shaft-Bipolar/254388269536?ssPageName=STRK%3AMEBIDX%3AIT&_trksid=p2060353.m2749.l2649
Driver
DM542 4.2A 18-50V
https://www.ebay.de/itm/1PC-Schrittmotortreiber-DM542-4-2A-18-50V-Nema17-Nema23-Stepper-MOTOR-Driver/254361335670?ssPageName=STRK%3AMEBIDX%3AIT&_trksid=p2060353.m2749.l2649
Encoder
OMRON 2000R (was in my stock, changend in the soft, a correct is orderd)

The pcb board is a veroboard






------ Originalnachricht ------
Von: "Maxim Kachurovskiy" <notifications@github.com>
An: "kachurovskiy/nanoels" <nanoels@noreply.github.com>
Cc: "Dreher-V8" <wolfgang.knoth@gmx.de>; "Comment" 
<comment@noreply.github.com>
Gesendet: 31.12.2020 14:45:07
Betreff: Re: [kachurovskiy/nanoels] Welcome to NanoEls Discussions! (#1)

**Reply by kachurovskiy** [reply]
Wolfgang, it sounds like your encoder might be connected to pins other than `D2` and `D3`.

Encoder pins should be D2 and D3 since Arduino Nano only supports interrupts on those 2 pins.

I've added some more info at the top of the page, hope it helps.

**Comment by Dreher-V8** [comment]
Maxim,
happy new year.

I found out, that Driver DM542 is not DM542. The "normal" (wo "y") can 
only >400 Steps, but the "y" Series can also 200 Steps.
I have ordert a new one as "y".
I will wait until both parts are here and give you then info

Best regards
Wolfgang


------ Originalnachricht ------
Von: "Maxim Kachurovskiy" <notifications@github.com>
An: "kachurovskiy/nanoels" <nanoels@noreply.github.com>
Cc: "Dreher-V8" <wolfgang.knoth@gmx.de>; "Comment" 
<comment@noreply.github.com>
Gesendet: 31.12.2020 17:08:19
Betreff: Re: [kachurovskiy/nanoels] Welcome to NanoEls Discussions! (#1)

**Reply by kachurovskiy** [reply]
Wolfgang, NanoEls can handle 400 steps, just adjust the corresponding constant in the code. Thanks!

**Comment by Dreher-V8** [comment]
Hallo Maxim,

I will test it in the next days.
Stepperonline can not sell the "y" Series, it is out of stock.
She send me an "T" series
The Encoder (600 steps) is also "on the fly"




------ Originalnachricht ------
Von: "Maxim Kachurovskiy" <notifications@github.com>
An: "kachurovskiy/nanoels" <nanoels@noreply.github.com>
Cc: "Dreher-V8" <wolfgang.knoth@gmx.de>; "Comment" 
<comment@noreply.github.com>
Gesendet: 08.01.2021 14:38:52
Betreff: Re: [kachurovskiy/nanoels] Welcome to NanoEls Discussions! (#1)

**Comment by Siroblong2021** [comment]
Hello Maxim, as i said i started my Nanoels Project. i have Spaceproblems mounting the Encoder on my lathe. I thought about a Disc on the rear of the spindle with holes and a laser counter. Is it possible without changing the code, how many holes would you say is necessary at least ?
I have seen other ELS Projects on YT that use much less than 600 Steps

**Reply by kachurovskiy** [reply]
Hi! Please attach some photos of your situation, hard to say what's going on without seeing how the parts fit (or don't fit) together! Best wishes.

**Reply by Siroblong2021** [reply]
Hello Maxim, 
after looking to other ELS Projects on YT, I decided to choose GT2  6 Millimeter Toothed Belt and two 100 Tooth Gears, one on the rear Mainspindle , one on the Encoder. I will send some Pics about the Progress of the Project, waiting for Parts from Ch.
PCB arrived today, already soldered. Testing Electronics with non atached Hardware starting soon.
Thank you very much for the publication of your Project and your Support, Stahlwurm

**Comment by Siroblong2021** [comment]
Hello Maxim, started testing today but i am still unexperienced in code.
Can you send me a code screenshot what you mean with "uncomment" for the BIG DISPLAY, how it has to look in the end?

Thanx a lot, Stahlwurm

**Reply by kachurovskiy** [reply]
Dear Stahlwurm, you need to replace line 32 in NanoEls.ino. By default it's "// #define LCD", make it "#define LCD". Cheers!

**Reply by Siroblong2021** [reply]
Hello Maxim, it worked instantly and flawless! 
Another Question: Since there is already an Encoder integrated in the Hardware and there is a free row left in the big Display would it be possible to adapt the code to show the mainspindle speed (Tachometer)? Greetings From Hanau,Germany

**Reply by kachurovskiy** [reply]
Hi Stahlwurm, updating the display takes Arduino 40ms and during this time the stepper can't be operated. This makes updating the display impossible during many normal turning operations even at medium RPM without making the stepper motion uneven. There exists plenty standalone RPM meters though e.g. https://www.aliexpress.com/item/32891009917.html which will operate completely independently from the NanoEls (which is good since those 2 things won't break at the same time) - cheers!

**Reply by Siroblong2021** [reply]
Hi Maxim, thats what i feared, it takes too long. 
So there will be a second Nano, next Project Tachometer. 
Is it possible to use the existing Encoder (Signal) for both, the ELS and the Tachometer?
Yes its easier to buy a Product but i am looking for Projects to improve my c++  and electronics skills 

Have a nice Evening, Stahlwurm

**Reply by kachurovskiy** [reply]
Stahlwurm, there's normally no need to use encoder to detect RPM. Just measuring the time of 1 full turn is enough. The system relies on a magnet glued to the main shaft and a magnetic sensor positioned to detect whenever a magnet passes by.

I realized that the AliExpress item above doesn't include the sensor/magnet - here's one that does - https://www.amazon.de/-/en/Digital-Tachometer-Sensor-Proximity-Switch/dp/B07G2HPK62 - cheers!

**Comment by R2-66** [comment]
Hi Maxim,

What an awesome project and thank you for sharing!

I've been digging into this project and very excited to get started however I wanted to pick your brain on making this work on an older lathe with a 16TPI leadscrew.

Based on the code (see below) do you think I could simply swap "#define LEAD_SCREW_HMM 200.0 // 2mm lead screw" to the conversion for 16TPI to "#define LEAD_SCREW_HMM 158.750 // 16TPI lead screw" ?

// Define your hardware parameters here. Don't remove the ".0" at the end.
#define ENCODER_STEPS 600.0 // 600 step spindle optical rotary encoder
#define MOTOR_STEPS 200.0 // 200 step stepper motor, no microstepping
#define LEAD_SCREW_HMM 200.0 // 2mm lead screw

Link to TPI to MM conversions - https://www.gewinde-normen.de/en/TPI.html

Again thank you and any help would be greatly appreciated.

Cheers,
George

**Reply by kachurovskiy** [reply]
George, thank you. Yes, I think that should work.

**Reply by R2-66** [reply]
Thanks Maxim, I'll give it a try and post progress as this project comes alive.

**Comment by Hedning003** [comment]
Hi Maxim .. have done the entire build,and it's now on my workbench-having an issue however, pcb fires up,lcddisplay working,buttons working,encoder works ..bought all stuff according to the "manual".
The nema23-motor wont turn when turning the encoder....have a +5v when turning the encoder counter clockwise,and zero when turning clockwize.
The display dont update movement until i stop spinning the encoder.
I have no volts on pu + and - at all.
the stepper shoud indicate if it recives signals,on one of the diodes-but nothing..
I have solder'd the 2 resistors to the pcb-pherhaps i shod remove them?
running the stepper at 40 volts,just to test.
Have no volts out to the motor when turning the encoder,or trying to just move left-right on the buttons..but indicates in the display movement

**Reply by Hedning003** [reply]
![ard1](https://user-images.githubusercontent.com/95235260/143897434-06ac8674-15e0-4ee3-b5ef-1d6a59584f84.jpg)
a pic. of the mess at the moment

**Reply by kachurovskiy** [reply]
Hi! Try testing the stepper separately using Arduino + sample sketches - see if your stepper driver and wiring is in order? Best wishes.

**Reply by Hedning003** [reply]
thanks for the comment
the stepper is brand new,and i have been over the wiring several times..have a cnc-mill and hooked it up the same way-arduino is something new to me,and i dont know what pins to activate to send signals to the stepper

**Reply by Hedning003** [reply]
Have no idea of what i did,or how..but it came alive😃😃 i'm a happy camper...i think it might have been due to me tinkering around,and then cleared the vram,cutting the power and switched it back on again.
Instalationpics. is comming in that thread

**Comment by Hedning003** [comment]
everythin hooked up on the lathe an played around today..i have a DRO on the lathe and i can se that the nanoels dosnt match the dro(not going back and forth...backlash)  going just one way and it  seems that the nanoels measure about 0.05mm to much for every 10mm of movement.
so my question is..can i simply fix this by playing thit the value in the .iso for the leadscrew pitch?  change from say 200.0 to 202.0 to get the els and the dro to measure the same?

**Reply by kachurovskiy** [reply]
Hedning, yes, adjusting the leadscrew pitch slightly in the code seems like the way to go. Thank you!

**Comment by ws786873** [comment]
I read your project. It removes complex gears and uses stepping motors to achieve different feed rates for turning threads. Is my understanding correct? This is really convenient for turning threads. I have a 3D printer, and I can easily buy materials and PCBs in China. I think I will try to make one. Thank you for your work~

**Reply by kachurovskiy** [reply]
Yes, your understanding is correct. Thank you!

**Comment by ws786873** [comment]
I don't know much about stepping motor and driver, so I want to buy a motor that can be used directly without modifying the software. So I want you to help me confirm whether the following motors are available. Rtelligent57cm31 has a size of 57.57.97, 3nm, 1.8 °, and a shaft diameter of 8mm (your shaft diameter doesn't seem to be described). This is a closed-loop motor. The corresponding driver is CL57. I really don't know what parameters I need to see about the starter to confirm whether it is available. Next is an encoder. Does it only need to be 600P? Does it need other parameters？

**Reply by kachurovskiy** [reply]
Hi ws786873, can you please post the picture of your lathe? For an average desktop lathe the motor you described is completely sufficient unless you need to perform heavy cuts e.g. take more than 1-2mm in one pass.

For the encoder, anything between 400 and 1000 steps should work. Best wishes!

**Reply by ws786873** [reply]
![2146157770](https://user-images.githubusercontent.com/25994407/145019177-96e7a5b5-d932-48e7-9fea-9bdacc1e62d6.jpg)
![401340340](https://user-images.githubusercontent.com/25994407/145019192-daea2b2b-b5ed-4ed5-afcb-655708136e5c.jpg)
![1566165432](https://user-images.githubusercontent.com/25994407/145019201-08ad3b0c-05af-4bc4-9402-716de33c18a9.jpg)
this is my  lathe,it's 210V.

**Reply by kachurovskiy** [reply]
For this lathe, the motor you cited should be strong enough.

**Reply by ws786873** [reply]
When I bought encoders, I found that they have NPN or NPN collector output, voltage output (NPN) and linear driver output. These four options. Their maximum response frequencies are 50Hz and 100Hz. The encoders I found seem to have three outputs of ABZ, while the encoder you use seems to have only AB output. Will there be a problem with the encoder using three-phase output。How should I buy it.

**Reply by kachurovskiy** [reply]
I'm no specialist on encoders, just a regular one like https://www.aliexpress.com/item/4000797890690.html should do it. 2 phases (AB), 5V, 600 ppr.

**Comment by ws786873** [comment]
<img width="343" alt="屏幕截图 2021-12-06 234211" src="https://user-images.githubusercontent.com/25994407/144876326-95e12454-ce39-4679-b81d-2c8c954225e8.png">
<img width="347" alt="屏幕截图 2021-12-06 234451" src="https://user-images.githubusercontent.com/25994407/144876753-d39689ff-4472-4f91-92b5-6d9ed74bebbd.png">

I modified the copper cladding and added the button resistance. But I don't know how to upload files.

**Reply by kachurovskiy** [reply]
Hi ws786873, you can share your files by dragging and dropping them into this comment field.

Thank you for sharing the PCB modification. Why do you think resistors are necessary for buttons?

**Reply by ws786873** [reply]
[PCBandSCH.zip](https://github.com/kachurovskiy/nanoels/files/7667834/PCBandSCH.zip)

**Reply by mockendon** [reply]
I could understand adding pull up or pull down resistors but these seem to be between the data pins and the switches - this might reduce the reliability, it certainly adds complexity to the board and assembly and is not standard practice for Arduino (see the reference designs on Arduino.cc .

**Comment by bankspa62** [comment]
Can any one help with the h5 ELS DRO. I installed the tft file to the nextion and that worked then I installed the ino file to the ESP32S3 board and all I get when booted up is the tft screen like there is no ino filed installed. Before I installed all of this I had just the ELS files on there and it worked fine.

**Reply by kachurovskiy** [reply]
> Before I installed all of this I had just the ELS files on there and it worked fine.

So the original version of H5 was working correctly and installing version from https://github.com/kachurovskiy/nanoels/discussions/367 caused the unresponsive screen?

Perhaps you need to compile and upload the modded version from Arduino IDE with additional settings like "Tools -> Erase All Flash Before Sketch Upload = Enabled".

**Reply by bankspa62** [reply]
Ok I verified all pin numbers again and enabled erase flash which it did do. software version is 97. Have right size screen for nextion but it still will not work. Even tried a new board. The only thing that it does do is when I touch the screen I can tell a small color change by pressing the buttons.

**Reply by bankspa62** [reply]
Something weird happen, I plugged my board with the H5 ino into the Nextion with the DRO screen and items on there started working.
So now I'm really confused on what to do. The mode selection and tpicth did not work.

**Reply by bankspa62** [reply]
I got it to work. After watching some videos I ran across one that showed hook the rx to the tx and tx to the rx, which by dumb luck I had it hooked that way on the H5 setup. This was the first time ever hooking up a nextion to a arduino. But I'm glad I ran across the video that show it. Thanks for your help.

**Reply by bankspa62** [reply]
I would like to know how to make the keypad, is there a video on it?

**Reply by kachurovskiy** [reply]
Glad you figured it out. Do you mean the DRO keypad from https://github.com/kachurovskiy/nanoels/discussions/367 or a regular PS2 keyboard?

**Reply by bankspa62** [reply]
The DRO Keypad

**Reply by kachurovskiy** [reply]
You're going to need to ask the author about that, it's not mine.

**Reply by bankspa62** [reply]
Thank you
